File: plugins/DEV-ESSENTIALS/hooks/doc-file-warning.js
#!/usr/bin/env node
/**
 * PreToolUse Hook: Doc File Warning
 *
 * Warns about creating non-standard .md/.txt files outside of
 * recognized locations (docs/, skills/, resources/).
 *
 * Non-blocking (exit code 0).
 */

const path = require('path');
const MAX_STDIN = 1024 * 1024;

const ALLOWED_NAMES = [
  'README', 'CLAUDE', 'CONTRIBUTING', 'CHANGELOG', 'LICENSE',
  'SKILL', 'AGENTS', 'INTERNALS', 'REFERENCE', 'TROUBLESHOOTING',
  'ADVANCED',
];

const ALLOWED_DIRS = ['docs', 'skills', 'resources', 'dev', 'commands', 'agents', '.cursor'];

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('doc-file-warning')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const filePath = input.tool_input?.file_path || '';

    if (!/\.(md|txt)$/i.test(filePath)) {
      process.stdout.write(data);
      return;
    }

    const baseName = path.basename(filePath, path.extname(filePath)).toUpperCase();
    if (ALLOWED_NAMES.includes(baseName)) {
      process.stdout.write(data);
      return;
    }

    const normalizedPath = filePath.replace(/\\/g, '/');
    const inAllowedDir = ALLOWED_DIRS.some(dir => normalizedPath.includes(`/${dir}/`) || normalizedPath.startsWith(`${dir}/`));
    if (inAllowedDir) {
      process.stdout.write(data);
      return;
    }

    console.error(`[Hook] WARNING: Creating non-standard doc file: ${filePath}`);
    console.error('[Hook] Consider placing documentation in docs/, skills/resources/, or using a standard name (README, CONTRIBUTING, etc.).');

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] doc-file-warning error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/DEV-ESSENTIALS/hooks/doc-file-warning.py
#!/usr/bin/env python3
"""
PreToolUse Hook: Doc File Warning

Warns about creating non-standard .md/.txt files outside of
recognized locations (docs/, skills/, resources/).

Non-blocking (exit code 0).
"""

import json
import os
import re
import sys

MAX_STDIN = 1024 * 1024

ALLOWED_NAMES = {
    "README", "CLAUDE", "CONTRIBUTING", "CHANGELOG", "LICENSE",
    "SKILL", "AGENTS", "INTERNALS", "REFERENCE", "TROUBLESHOOTING",
    "ADVANCED",
}

ALLOWED_DIRS = ["docs", "skills", "resources", "dev", "commands", "agents", ".cursor"]


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "doc-file-warning" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        file_path = inp.get("tool_input", {}).get("file_path", "")

        if not re.search(r"\.(md|txt)$", file_path, re.IGNORECASE):
            sys.stdout.write(data)
            return

        base_name = os.path.splitext(os.path.basename(file_path))[0].upper()
        if base_name in ALLOWED_NAMES:
            sys.stdout.write(data)
            return

        normalized = file_path.replace("\\", "/")
        if any(f"/{d}/" in normalized or normalized.startswith(f"{d}/") for d in ALLOWED_DIRS):
            sys.stdout.write(data)
            return

        print(f"[Hook] WARNING: Creating non-standard doc file: {file_path}", file=sys.stderr)
        print("[Hook] Consider placing documentation in docs/, skills/resources/, or using a standard name (README, CONTRIBUTING, etc.).", file=sys.stderr)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] doc-file-warning error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
